src/Core/Grid/Grid.php

```diff
@@ -0,0 +1,266 @@
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid;
+
+use PrestaShop\PrestaShop\Core\Grid\Action\RowActionCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnCollectionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnInterface;
+use PrestaShop\PrestaShop\Core\Grid\Data\GridDataInterface;
+use PrestaShop\PrestaShop\Core\Grid\Definition\GridDefinitionInterface;
+use PrestaShop\PrestaShop\Core\Grid\Exception\MissingColumnInRowException;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use PrestaShop\PrestaShop\Core\Grid\View\GridView;
+use Symfony\Component\Form\FormInterface;
+
+/**
+ * Class Grid is responsible for holding final Grid data
+ */
+final class Grid
+{
+    /**
+     * @var GridDefinitionInterface
+     */
+    private $definition;
+
+    /**
+     * @var GridDataInterface
+     */
+    private $data;
+
+    /**
+     * @var SearchCriteriaInterface
+     */
+    private $searchCriteria;
+
+    /**
+     * @var FormInterface
+     */
+    private $filterForm;
+
+    public function __construct(
+        GridDefinitionInterface $definition,
+        GridDataInterface $data,
+        SearchCriteriaInterface $searchCriteria,
+        FormInterface $filterForm
+    ) {
+        $this->definition = $definition;
+        $this->data = $data;
+        $this->searchCriteria = $searchCriteria;
+        $this->filterForm = $filterForm;
+    }
+
+    /**
+     * @return GridView
+     */
+    public function createView()
+    {
+        $view = new GridView(
+            $this->definition->getIdentifier(),
+            $this->definition->getName(),
+            $this->createColumnsView(),
+            $this->filterForm->createView()
+        );
+
+        $view->setBulkActions($this->createBulkActionsView());
+        $view->setGridActions($this->createGridActionsView());
+        $view->setData([
+            'rows' => $this->createRowsView(),
+            'rows_total' => $this->data->getRowsTotal(),
+            'query' => $this->data->getQuery(),
+        ]);
+        $view->setPagination([
+            'offset' => $this->searchCriteria->getOffset(),
+            'limit' => $this->searchCriteria->getLimit(),
+        ]);
+        $view->setSorting([
+            'order_by' => $this->searchCriteria->getOrderBy(),
+            'order_way' => $this->searchCriteria->getOrderWay(),
+        ]);
+
+        return $view;
+    }
+
+    /**
+     * Get rows data ready for rendering in template
+     *
+     * @return array
```

Comment:
we also need a RowsViewsCollection here (sorry).

So you will extract the construction of the array into RowsViewsCollection::fromArray(array $rowsView).

And this function will end this way:

```php
return RowsViewsCollection::fromArray($rowsView);
```